pkg/virt-api/rest/subresource.go

```diff
@@ -180,7 +205,66 @@ func (app *SubresourceAPIApp) VNCRequestHandler(request *restful.Request, respon
 	getConsoleURL := func(vmi *v1.VirtualMachineInstance, conn kubecli.VirtHandlerConn) (string, error) {
 		return conn.VNCURI(vmi)
 	}
-	app.streamRequestHandler(request, response, validate, getConsoleURL)
+	app.streamRequestHandler(request, response, validate, app.virtHandlerDialer(getConsoleURL), rawStreamers)
+}
+
+func (app *SubresourceAPIApp) TunnelRequestHandler(request *restful.Request, response *restful.Response) {
+	validate := func(vmi *v1.VirtualMachineInstance) *errors.StatusError {
+		condManager := controller.NewVirtualMachineInstanceConditionManager()
+		if condManager.HasCondition(vmi, v1.VirtualMachineInstancePaused) {
+			return errors.NewConflict(v1.Resource("virtualmachineinstance"), vmi.Name, fmt.Errorf("VMI is paused"))
+		}
+		return nil
+	}
+	app.streamRequestHandler(request, response, validate, tcpDialer(request), websocketStreamers)
+}
+
+func tcpDialer(request *restful.Request) dialer {
+	return func(vmi *v1.VirtualMachineInstance) (net.Conn, *errors.StatusError) {
+		logger := log.Log.Object(vmi)
+
+		targetIP, err := getTargetInterfaceIP(vmi, request.QueryParameter("networkInterface"))
```

Comment:
we might not be guaranteed that the virt-api server has connectivity to all VMI interfaces. I think it's technically possible for a VMI to have an interface that isn't reachable from the pod network. 

So we might need to restrict this to only the pod network interface.

Reply:
Yeah, I think so too.
That's why I added a way to specify it on the backend already. The first network might not be the right one.